smithy-codegen-core/src/test/java/software/amazon/smithy/codegen/core/TraceFileTest.java

```diff
@@ -0,0 +1,311 @@
+                .addTag("requestBuilder", "AWS SDK request builder")
+                .addType("TYPE", "Class, interface (including annotation type), or enum declaration")
+                .build();
+
+        /**
+         * Building TraceFile - build the trace file by passing the different objects to the builder
+         * You can either construct the shapes map yourself, and add it to TraceFile, or you can add each shape
+         * individually as shown below.
+         * SmithyTrace is a constant in my code set to 1.0 for this version, so you don't have to worry about
+         * setting it.
+         */
+        TraceFile.Builder traceFileBuilder = TraceFile.builder()
+                .artifact(artifactMetadata)
+                .definitions(artifactDefinitions);
+
+        //adding one ShapeLink to TraceFile
+        type = "TYPE";
+        id = "software.amazon.awssdk.services.snowball.SnowballClient";
+        traceFileBuilder.addShapeLink("com.amazonaws.snowball#Snowball",
+                ShapeLink.builder()
+                        .type(type)
+                        .id(id)
+                        .addTag("service")
+                        .build());
+
+        //adding multiple ShapeLinks for the same ShapeId; can also add a List<ShapeLink>
+        type = "TYPE";
+        id = "software.amazon.awssdk.services.snowball.model.ListClusterJobsRequest$Builder";
+        traceFileBuilder.addShapeLink("com.amazonaws.snowball#ListClustersRequest",
+                ShapeLink.builder()
+                        .type(type)
+                        .id(id)
+                        .addTag("requestBuilder")
+                        .build());
+        type = "TYPE";
+        id = "software.amazon.awssdk.services.snowball.model.ListClusterJobsRequest";
+        traceFileBuilder.addShapeLink("com.amazonaws.snowball#ListClustersRequest",
+                ShapeLink.builder()
+                        .type(type)
+                        .id(id)
+                        .addTag("request")
+                        .build());
+
+        //finally, build the TraceFile
+        TraceFile traceFile = traceFileBuilder.build();
+
+        /**
+         * After creating a TraceFile, you may want to validate it.
+         * 1) You can validate whether all your types/tags in ShapeLink are in the definitions object
+         * 2) You can validate whether all your TraceFile matches your model by checking if all the ShapeIds
+         * in a model file are in the TraceFile and all the ShapeIds in a model file are in the TraceFile
+         */
+
+        //types/tags validation
+        traceFile.validateTypesAndTags();
+
+        /**
+         * model validation, see the correctValidateModel test for another example, I don't have the actual model for
+         * this trace file, so cannot validate it.
+         */
+        //traceFile.validateModel(assembleModelTestHelper(("Your model resource path here");
```

Comment:
Should this be uncommented?

Reply:
No, because we do not have a Smithy model for the TraceFile that is being created in this example. I added a comment that explains this and points users to look at the actual model validation tests to see a complete example. The purpose of including this here was to give a more complete example test for Ali to show him what he would need to do/anyone creating a TraceFile from scratch.